Debounce/Bufferer.cs

```diff
@@ -0,0 +1,119 @@
+﻿// SPDX-FileCopyrightText: 2024 Alain van den Berg
+//
+// SPDX-License-Identifier: MIT
+
+namespace Dorssel.Utilities;
+
+/// <summary>
+/// A <see cref="Debouncer"/> that buffers events of type <typeparamref name="TEvent"/> that were triggered, aggregates them and sends aggregated events.
+/// </summary>
+/// <typeparam name="TEvent">Events to buffer</typeparam>
+/// <remarks>
+/// This is not as performant as the <see cref="Debouncer"/> due to allocations.
+/// </remarks>
+public sealed class Bufferer<TEvent> : IDisposable, IBufferEvents<TEvent>
+{
+    IDebounceEvents debouncer;
+    List<TEvent> eventList = new();
+    object eventListLock = new();
+
+    /// <summary>
+    /// Create Bufferer with the default <see cref="Debouncer"/>.
+    /// </summary>
+#pragma warning disable CA2000 // Dispose objects before losing scope
+    public Bufferer() : this(new Debouncer()) { }
+#pragma warning restore CA2000 // Dispose objects before losing scope
+
+    /// <summary>
+    /// Create Bufferer with a specific <see cref="IDebounceEvents"/> instance.
+    /// </summary>
+    /// <param name="debouncer"></param>
+    /// <exception cref="ArgumentNullException"></exception>
+    public Bufferer(IDebounceEvents debouncer)
+    {
+        this.debouncer = debouncer ?? throw new ArgumentNullException(nameof(debouncer));
+        debouncer.Debounced += Debouncer_Debounced;
+    }
+
+    /// <summary>
+    /// Wrap the Debounced event and invoke the Buffered event instead.
+    /// </summary>
+    private void Debouncer_Debounced(object sender, DebouncedEventArgs e)
+    {
+        List<TEvent> debouncedEvents;
+        lock (eventListLock)
+        {
+            debouncedEvents = eventList;
```

Comment:
This implicitly casts it to a IReadOnlyList, but consumers can still cast it back to IList and modify it. Using https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.asreadonly?view=net-8.0 you can make a true readonly list. At the cost of the allocation of the wrapper (it doesn't actually copy the list, just wraps it in a class instance that *only* implements IReadOnlyList. The cost of the allocation is acceptable, given that Bufferer alread needs allocation for all the events.

Reply:
Good point, although a bit convoluted. Since `AsReadOnly()` is not supported for netstandard1.2, I opted for `new ReadOnlyCollection()`.

Reply:
That is, in fact, exactly how `AsReadOnly` is implemented, see https://github.com/dotnet/runtime/blob/d88c7ba88627b4b68ad523ba27cb354809eb7e67/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L277-L278

Convoluted, yes, maybe. Now you cannot modify the list, but you can still modify the data *in* the list. But we are not going to make deep copies for every listener...

In any case, I like this a little better.